### PR TITLE
Igvf 45 add donor

### PR DESCRIPTION
**Pull request by KAndreeva**
Implementation of Donor abstract class and human_donor, and rodent_donor concrete classes.
Encountered an issue with array of enums being enforced by json schema validation.
Biosample schema updated to link to Donor.

### CODE REVIEW

File: src/igvfd/tests/data/inserts/human_donor.json [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+            {
+                "health_description": "Running fever 103F, feeling tired",
+                "date_start": "2021-12-03",
+                "date_end": "2022-01-10"
+            }
+        ],
+        "ethnicity": [
+            "Hispanic"
+        ],
+        "genetic_modifications": [
```
Comment (cricketsloan) [inline src/igvfd/tests/data/inserts/human_donor.json:29]: does it make sense to have a genetic modification on a human.  In general, our notes said that we were not going to include this on donor.

File: src/igvfd/tests/test_schema_human_donor.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        human_donor['@id'],
+        {'ethnicity': ['Elf']}, expect_errors=True)
+    print('status code: ' + str(res.status_code))
+    assert(res.status_code == 422)
+
+
+def test_donor_with_parents(human_donor, parent_human_donor_1, parent_human_donor_2, testapp):
+    res = testapp.patch_json(
+        human_donor['@id'],
+        {'parents': [parent_human_donor_1['@id'], parent_human_donor_2['@id']]})
```
Comment (cricketsloan) [inline src/igvfd/tests/test_schema_human_donor.py:37]: I think that this test would have been more efficiently tested with the use of inserts that have parents.

File: src/igvfd/tests/test_schema_human_donor.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        {'parents': []})
+    assert(res.status_code == 200)
+
+
+def test_collections(human_donor, testapp):
+    res = testapp.patch_json(
+        human_donor['@id'],
+        {'collections': ['ENCODE']})
+    assert(res.status_code == 200)
+    # this is working?  Unknown why enum values in array not enforced.
```
Comment (cricketsloan) [inline src/igvfd/tests/test_schema_human_donor.py:53]: This commented out code should be removed.  Did it get resolved?

File: src/igvfd/tests/test_schema_rodent_donor.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            }
+        ]})
+    assert(res.status_code == 200)
+
+
+def test_donor_with_three_parents(rodent_donor, parent_rodent_donor1, parent_rodent_donor2, parent_rodent_donor3, testapp):
+    res = testapp.patch_json(
+        rodent_donor['@id'],
+        {'parents': [
+            parent_rodent_donor1['@id'],
```
Comment (cricketsloan) [inline src/igvfd/tests/test_schema_rodent_donor.py:33]: What is a rodent with 3 parents representing?  Can we limit the array to length 2?

**Review by cricketsloan — CHANGES_REQUESTED** [review]
This is looking good.  There were a few things that I was unclear on that I left comments in the text.  There is an overall question that I have about the external_resources.  In the past this has been a link to an entry in another database that represents the donor so like GEO or MGI.  I am not certain what it is representing now.

File: src/igvfd/schemas/human_donor.json [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+        },
+        "ethnicity": {
+            "title": "Ethnicity",
+            "comment": "We use http://bioportal.bioontology.org/ontologies/HANCESTRO as a template for the enum list of ethnicities.",
+            "type": "array",
+            "uniqueItems": true,
+            "minItems": 1,
+            "items": {
+                "type": "string",
+                "enum": [
```
[the remark this answers is not in the document]
Comment (gabdank) [reply src/igvfd/schemas/human_donor.json:125]: yes, it ios from ENCODE human donor:  "enum": [
            "African American",
            "Arab",
            "Asian",
            "Black",
            "Black African",
            "Chinese",
            "Esan",
            "Eskimo",
            "European",
            "Gambian",
            "Han Chinese",
            "Hispanic",
            "Indian",
            "Japanese",
            "Luhya",
            "Maasai",
            "Mende",
            "Native Hawaiian",
            "Yoruba"
        ]

File: src/igvfd/schemas/mixins.json [hunk 6]
```diff
@@ -299,5 +299,35 @@
...
+            "uniqueItems": true,
+            "items": {
+                "title": "Genetic modification",
+                "description": "Genetic modification(s).",
+                "type": "string",
+                "$comment": "This should later become a linkTo for GeneticModifications"
+            }
+        }
+    },
+    "references": {
```
[the remark this answers is not in the document]
Comment (gabdank) [reply src/igvfd/schemas/mixins.json:318]: @KAndreeva yes, we decided to NOT link from donor to publication object. just add regular expression to make sure it is doi or pubmed

File: src/igvfd/tests/data/inserts/human_donor.json (hunk 1 above)
Comment (gabdank) [reply src/igvfd/tests/data/inserts/human_donor.json:29]: drop genetic modification

File: src/igvfd/schemas/donor.json [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+                        "resource_url": {
+                            "title": "External Resource URL",
+                            "description": "URL for the external resource",
+                            "type": "string"
+                        }
+                    }
+                }
+            }
+        },
+        "collections": {
```
[the remark this answers is not in the document]
Comment (KAndreeva) [reply src/igvfd/schemas/donor.json]: Done

File: src/igvfd/schemas/human_donor.json [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+            "default": [],
+            "items": {
+                "title": "Health Status",
+                "type": "object",
+                "additionalProperties": false,
+                "properties": {
+                    "health_description": {
+                        "title": "Description of Donor Health",
+                        "description": "Description of the health status change.",
+                        "type": "string",
```
[the remark this answers is not in the document]
Comment (KAndreeva) [reply src/igvfd/schemas/human_donor.json:87]: I believe it is too early to restrict this field.  My intent was for this to be just a health description of the donor's condition.

File: src/igvfd/schemas/mixins.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (KAndreeva) [reply src/igvfd/schemas/mixins.json:335]: Done

File: src/igvfd/tests/data/inserts/human_donor.json (hunk 1 above)
Comment (KAndreeva) [reply src/igvfd/tests/data/inserts/human_donor.json]: Done

File: src/igvfd/tests/test_schema_rodent_donor.py (hunk 4 above)
Comment (KAndreeva) [reply src/igvfd/tests/test_schema_rodent_donor.py:33]: Done

Comment (KAndreeva) [reply src/igvfd/tests/test_schema_rodent_donor.py:33]: I have added a unit test for having a limit of two parents for human_donor and rodent_donor.

File: src/igvfd/tests/test_schema_human_donor.py (hunk 2 above)
Comment (KAndreeva) [reply src/igvfd/tests/test_schema_human_donor.py:37]: This is tested in both places.   I set parents in the insert test data.

File: src/igvfd/schemas/donor.json [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+        },
+        "sex": {
+            "title": "Sex",
+            "description": "Sex of the donor.",
+            "type": "string",
+            "enum": [
+                "male",
+                "female"
+            ]
+        },
```
Comment (gabdank) [inline src/igvfd/schemas/donor.json:73]: do we want "unknown"?

File: src/igvfd/schemas/rodent_donor.json [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+            "$ref": "mixins.json#/url"
+        },
+        {
+            "$ref": "mixins.json#/references"
+        },
+        {
+            "$ref": "mixins.json#/collections"
+        }
+    ],
+    "dependentSchemas": {
```
Comment (gabdank) [inline src/igvfd/schemas/rodent_donor.json:69]: can not we just $merge all the properties, not only dependencies?

File: src/igvfd/tests/test_schema_human_donor.py (hunk 2 above)
Comment (gabdank) [reply src/igvfd/tests/test_schema_human_donor.py:37]: I am not following the comment or the answer, can you please clarify?

Comment (gabdank) [reply src/igvfd/tests/test_schema_human_donor.py:37]: I think now I understand. We would drop the tests for no parents and 2 parents, leaving these for inserts, but will leave the test that checks that 3 parents are 422.

File: src/igvfd/schemas/donor.json [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+            "$ref": "mixins.json#/url"
+        },
+        {
+            "$ref": "mixins.json#/references"
+        },
+        {
+            "$ref": "mixins.json#/collections"
+        }
+    ],
+    "dependentSchemas": {},
```
Comment (gabdank) [inline src/igvfd/schemas/donor.json:57]: we should not include empty dependencies

**Review by gabdank — APPROVED** [review]

**Pull request merged**